docs/convenience.rst

```diff
@@ -63,119 +61,63 @@ overcome transient failures.  This typically looks something like:
         else:
            break
 
-This is rather ugly.
+This is rather ugly and easy to get wrong.
 
-Transaction managers provide a helper for this case. To show this,
-we'll use a contrived example:
+Transaction managers provide two helpers for this case.
 
-.. doctest::
+Running and retrying functions as transactions
+______________________________________________
 
-    >>> ntry = 0
-    >>> with transaction.manager:
-    ...      dm['ntry'] = 0
-
-    >>> import transaction.interfaces
-    >>> class Retry(transaction.interfaces.TransientError):
-    ...     pass
-
-    >>> for attempt in transaction.manager.attempts():
-    ...     with attempt as t:
-    ...         t.note('test')
-    ...         print("%s %s" % (dm['ntry'], ntry))
-    ...         ntry += 1
-    ...         dm['ntry'] = ntry
-    ...         if ntry % 3:
-    ...             raise Retry(ntry)
-    0 0
-    0 1
-    0 2
-
-The raising of a subclass of TransientError is critical here. It's
-what signals that the transaction should be retried.  It is generally
-up to the data manager to signal that a transaction should try again
-by raising a subclass of TransientError (or TransientError itself, of
-course).
-
-You shouldn't make any assumptions about the object returned by the
-iterator.  (It isn't a transaction or transaction manager, as far as
-you know. :)  If you use the ``as`` keyword in the ``with`` statement,
-a transaction object will be assigned to the variable named.
-
-By default, it tries 3 times. We can tell it how many times to try:
+The first helper runs a function as a transaction::
 
-.. doctest::
+    def do_somthing():
+        "Do something"
+        ... some something ...
 
-    >>> for attempt in transaction.manager.attempts(2):
-    ...     with attempt:
-    ...         ntry += 1
-    ...         if ntry % 3:
-    ...             raise Retry(ntry)
-    Traceback (most recent call last):
-    ...
-    Retry: 5
+    transaction.manager.run(do_somthing)
 
-It it doesn't succeed in that many times, the exception will be
-propagated.
+Of course you can run this as a decorator::
 
-Of course, other errors are propagated directly:
+    @transaction.manager.run
+    def do_somthing():
```

Comment:
Personally, I would drop this section, and simply leave the example at `transaction.manager.run(do_something)`. 

It's fine if `run` can be used as a decorator for those that want to, but I don't feel like it's something we should encourage. That's so rarely done, in comparison to how decorators are normally used, that it is unexpected and surprising; most people won't be expecting the instant execution and will be confused upon reading the code. 

I know that came up on the mailing list, but even "run" to me doesn't sufficiently imply the concept of "no, really, this is not a normal decorator, this is happening now."

Another thing I dislike about the use of decorators to actually run a function is that (once you understand that's what's happening) you have to read the function body, and *then* go back up and remember that it's actually being done now. It's backwards (or upside down?) and less natural to read than definition followed by execution.

It has the disadvantage of not naturally returning a result, too. Or rather, yes, it does return a result, but the result isn't the function as one would expect from the `def` statement. This, to me, is very surprising code:

```python
@transaction.manager.run
def create_user():
   ...
   return some_user if user_created else None

if create_user is not None: 
   # What? How could it be None, we just def'd it
```

If you try to use this pattern, you run into a conflict: naming the function such that it's meaningful in the transaction note, and naming the function such that it describes what it's actually doing, conflicts with naming the function such that it makes a good result variable name. Wouldn't this be more clear:
```python
def create_user():
   ...
   return some_user if user_created else None

user = transaction.manager.run(create_user)
if user is not None:
   ...
```

Finally, as a minor point, if `run` *isn't* usable as a decorator, it may make it somewhat easier to support passing arguments to the function, which in turn supports defining functions only once (not at execution time), which can improve code sharing, decrease execution time and reduce garbage collection (inner functions tend to lead to cycles unless you're extremely careful).

```python
user = transaction.manager.run(create_user, username='foo')
```